napari/_vispy/overlays/interaction_box.py

```diff
@@ -0,0 +1,375 @@
+    BC = boxes[:, 6] - boxes[:, 4]
+    BM = boxes[:, 6]
+
+    ABAM = np.multiply(AB, AM).sum(1)
+    ABAB = np.multiply(AB, AB).sum(1)
+    BCBM = np.multiply(BC, BM).sum(1)
+    BCBC = np.multiply(BC, BC).sum(1)
+
+    c1 = 0 <= ABAM
+    c2 = ABAM <= ABAB
+    c3 = 0 <= BCBM
+    c4 = BCBM <= BCBC
+
+    inside = np.all(np.array([c1, c2, c3, c4]), axis=0)
+
+    return inside
+
+
+class VispyInteractionBox:
+    def __init__(self, viewer, parent=None, order=0):
+
+        self._viewer = viewer
+        self._interaction_box = viewer.overlays.interaction_box
+        self.node = Compound([Line(), Markers()], parent=parent)
+        self.node.order = order
+        self._on_interaction_box_change()
+        self.initialize_mouse_events(viewer)
+        self._interaction_box.events.points.connect(
+            self._on_interaction_box_change
+        )
+        self._interaction_box.events.show.connect(
+            self._on_interaction_box_change
+        )
+        self._interaction_box.events.show_handle.connect(
+            self._on_interaction_box_change
+        )
+        self._interaction_box.events.show_vertices.connect(
+            self._on_interaction_box_change
+        )
+        self._interaction_box.events.transform.connect(
+            self._on_interaction_box_change
+        )
+        self._highlight_width = 1.5
+        self._selected_vertex: int = None
+        self._fixed_vertex: int = None
+        self._fixed_aspect: float = None
+        self._vertex_size = 10
+
+        self._highlight_color = (0, 0.6, 1)
+
+    @property
+    def marker_node(self):
+        """sequence of float: Scale factors."""
+        return self.node._subvisuals[1]
+
+    @property
+    def line_node(self):
+        """sequence of float: Scale factors."""
+        return self.node._subvisuals[0]
+
+    def _on_interaction_box_change(self, event=None):
+        """Called whenever the interaction box changed."""
+
+        # Compute the location and properties of the vertices and box that
+        # need to get rendered
+        (
+            vertices,
+            face_color,
+            edge_color,
+            pos,
+            width,
+        ) = self._compute_vertices_and_box()
+
+        if vertices is None or len(vertices) == 0:
+            vertices = np.zeros((1, self._viewer.dims.ndisplay))
+            size = 0
+        else:
+            vertices = vertices + 0.5
+            size = 10
+
+        self.marker_node.set_data(
+            vertices,
+            size=size,
+            face_color=face_color,
+            edge_color=edge_color,
+            edge_width=1.5,
+            symbol='square',
+            scaling=False,
+        )
+
+        if pos is None or len(pos) == 0:
+            pos = np.zeros((1, self._viewer.dims.ndisplay))
+            width = 0
+        else:
+            pos = pos + 0.5
+        self.line_node.set_data(pos=pos, color=edge_color, width=width)
+
+    def initialize_mouse_events(self, layer):
```

Comment:
Do we need to add the mouse callbacks to the layer? Can we have them go on the viewer? Maybe can they be defined alongside the interaction box model? I don't love the vispy object getting so complicated. In general we've tried to define mouse functions along side the models and not inside the vispy objects

Reply:
Sorry, they already were on the viewer, I just did not rename the argument (fixed).

The problem with having the mouse events in the model is that the model than needs to be aware of the viewer object. My impression, mainly reading #1325, was that the models should remain very simple and unaware of implementation detail. I agree that there is now a lot of code in the vispy overlay. Maybe I could split it in a "Viewer" component that only handles rendering and a "Controller" component that handles the mouse events?

Reply:
Also, is there a way to remove callbacks from the callback list? There is no reason why the InteractionBox should be listening to events when it is not active. Right now the callback will just terminate immediately when not active.

Reply:
hmm yeah, we don't want this to have to know about the viewer. looking through code here it looks like the only viewer attribute that gets read is `viewer.dims.ndisplay`, so maybe there's a way to just pass that information and then you don't need to pass around the viewer as much? Maybe @andy-sweet can advise here too (though I know he's a little less familiar with this part of the code too). Otherwise I can look in more detail later maybe

Reply:
The viewer.dims.ndisplay is not really that essential, because for now this only works in 2D anyway. The main issue is appending the callbacks to the mouse-callback lists of the viewer object.

Reply:
> The main issue is appending the callbacks to the mouse-callback lists of the viewer object.

Ok yeah, we should be able to figure out a different way to do that, so the logic of adding/ removing the callback happens inside the viewer, but the callback itself is defined in a file like `_interaction_box_mouse_bindings.py`. Maybe easier said then done, but that would be ideal I think

Reply:
Ok, I moved the mouse binding code into a class called InteractionBoxMouseBindings that lives in `_interaction_box_mouse_bindings.py`. It gets instanced in `qt_viewer.py` together with the vispy class.